modules/ingester/flush.go

```diff
@@ -165,49 +168,64 @@ func (i *Ingester) flushLoop(j int) {
 			return
 		}
 		op := o.(*flushOp)
+		op.attempts++
+
+		retry := false
 
-		var completeBlockID uuid.UUID
-		var err error
 		if op.kind == opKindComplete {
 			level.Debug(log.Logger).Log("msg", "completing block", "userid", op.userID)
 			instance, exists := i.getInstanceByID(op.userID)
 			if !exists {
 				// instance no longer exists? that's bad, log and continue
-				level.Error(log.Logger).Log("msg", "instance not found", "tenantID", op.userID)
+				level.Error(log.Logger).Log("msg", "instance not found", "userID", op.userID, "block", op.blockID.String())
 				continue
 			}
 
-			completeBlockID, err = instance.CompleteBlock(op.blockID)
-			if completeBlockID != uuid.Nil {
+			err := instance.CompleteBlock(op.blockID)
+			if err != nil {
+				handleFlushError(op, err)
+
+				if op.attempts >= maxCompleteAttempts {
+					level.Error(log.WithUserID(op.userID, log.Logger)).Log("msg", "Block exceeded max completion errors. Deleting. POSSIBLE DATA LOSS",
+						"userID", op.userID, "attempts", op.attempts, "block", op.blockID.String())
+					instance.ClearCompletingBlock(op.blockID)
+				} else {
+					retry = true
+				}
+			} else {
 				// add a flushOp for the block we just completed
 				i.flushQueues.Enqueue(&flushOp{
 					kind:    opKindFlush,
-					from:    time.Now().Unix(),
+					at:      time.Now(),
 					userID:  instance.instanceID,
-					blockID: completeBlockID,
+					blockID: op.blockID,
 				})
 			}
 
 		} else {
-			level.Debug(log.Logger).Log("msg", "flushing block", "userid", op.userID, "fp")
+			level.Info(log.Logger).Log("msg", "flushing block", "userid", op.userID, "block", op.blockID.String())
```

Comment:
i what you've done is the right choice, but am wondering if we will eventually want max attempts here too.